### PR TITLE
Add knowledge-ingest tool for URL content extraction

### LINKED ISSUE #76: Add knowledge-ingest tool for URL content extraction
## Overview

Add a new MCP tool `knowledge-ingest` that accepts a URL, fetches and extracts clean text content, and returns it to the calling agent for note creation via `knowledge-store`.

## Motivation

Currently, agents must rely on their own web-fetching capabilities (which vary wildly across clients) to capture external content. A server-side extraction tool provides:
- **Deterministic, consistent** content extraction regardless of calling model
- Clean markdown output stripped of nav, ads, boilerplate
- A reliable pipeline: `knowledge-ingest` → agent summarizes → `knowledge-store`

This is the biggest gap vs. Karpathy's "LLM Wiki" pattern, where URL ingestion is a first-class operation that feeds the knowledge base.

### Deterministic vs. Probabilistic Boundary

| Aspect | Layer | Control |
|--------|-------|---------|
| URL fetch + HTML parsing | Server-side | 🟢 Deterministic |
| Content extraction (readability) | Server-side | 🟢 Deterministic |
| Markdown conversion | Server-side | 🟢 Deterministic |
| Summarization / note creation | Calling agent | 🔴 Probabilistic (already is — no change) |

The tool does NOT create notes or summarize. It extracts and returns clean content. The agent then calls `knowledge-store` as it normally would. This keeps the probabilistic layer where it already lives.

### Tool Interface

```
knowledge-ingest
├── Input: { url: string }
├── Output: { title, content, url, extracted_at, word_count }
└── Errors: invalid URL, fetch failed, extraction failed
```

### Implementation Approach

- Use a readability library (e.g., `@mozilla/readability` + `linkedom` for DOM parsing without a browser) to extract article content
- Convert to clean markdown
- Return structured result with metadata
- No dependencies on any LLM — pure server-side extraction

## Out of Scope

- Auto-creating notes (agent decides what to store)
- Auto-cross-referencing (handled by separate enhancement)
- PDF/video extraction (future extension)
- Caching/deduplication of URLs

### PR DESCRIPTION
**Pull request by mrosnerr**
## Summary

New MCP tool `knowledge-ingest` that fetches a URL and extracts its article content as clean markdown. Fully deterministic — no LLM dependency. Agents use the extracted content to create notes via `knowledge-store`.

## Architecture

```
Agent calls knowledge-ingest(url) 
  → fetchHtml (Bun fetch + timeout + size guard)
  → extractArticle (linkedom DOM parsing + Mozilla Readability)  
  → turndown (HTML → clean markdown)
  → returns { title, content, wordCount, byline, excerpt, siteName, extractedAt }
```

## New Dependencies

| Package | Version | Purpose | Size |
|---------|---------|---------|------|
| `@mozilla/readability` | 0.6.0 | Article extraction (Firefox Reader View engine) |  ~30KB |
| `linkedom` | 0.18.12 | Lightweight DOM parser for server-side HTML |  ~30KB |
| `turndown` | 7.2.4 | HTML → Markdown conversion |  ~20KB |
| `@types/turndown` | 5.0.6 | TypeScript types (devDep) | — |

All pure JS, zero native deps, Bun-compatible. Same stack used by RSSNext/Folo in production.

## Changes

- **`src/url-extractor.ts`** (new): `isValidUrl()`, `fetchHtml()`, `extractArticle()`, `extractFromUrl()` 
- **`src/tool-handlers.ts`**: Added `IngestArgs` interface and `handleIngest()` handler
- **`src/mcp-server.ts`**: Registered `knowledge-ingest` tool with Zod schema
- **`tests/url-extractor.test.ts`** (new): 36 tests covering:
  - URL validation (10 tests): http/https/ftp/javascript/file/data/empty/paths/params
  - Article extraction (15 tests): title, markdown conversion, nav stripping, headings, lists, links, word count, byline, excerpt, empty/nav-only HTML
  - Fetch error handling (5 tests): HTTP errors, content-type validation, size limits
  - Integration (2 tests): handleIngest error propagation
  - Edge cases (4 tests): empty HTML, non-article pages, multi-article pages

## Non-Breaking

- New tool — no changes to existing tool behavior
- All 505 existing tests pass
- 0 lint errors

## Security

- URL validation: only `http:` and `https:` protocols allowed
- Content size limit: 5MB max HTML
- Fetch timeout: 15s default
- No code execution — pure content extraction

Closes #76

## Summary by CodeRabbit

* **New Features**
  * Added a knowledge-ingest tool to extract article content from URLs or provided HTML, returning structured markdown and metadata (title, byline, excerpt, site, word count)
  * Added markdown content-splitting and link-extraction utilities to produce sections and canonicalized links
  * Added repository search by URL to locate related notes

* **Documentation**
  * Updated tools reference, README, and skill guide to describe the new ingestion flow and limitations

* **Tests**
  * Comprehensive test suites for URL extraction, fetching, SSRF protections, redirects, and markdown processing

* **Chores**
  * Added HTML parsing and markdown-conversion dependencies and type definitions

### CODE REVIEW

File: src/url-extractor.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+  const controller = new AbortController();
+  const timer = setTimeout(() => controller.abort(), timeout);
+
+  try {
+    const response = await fetch(url, {
+      signal: controller.signal,
+      headers: { 'User-Agent': USER_AGENT },
+      redirect: 'follow',
+    });
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts]: Fixed in 8111c5f. Implemented all three mitigations:

1. `redirect: 'manual'` with manual redirect following (`MAX_REDIRECTS=5`)
2. `isPrivateOrReservedHost()` validates each redirect target — blocks 127.x, 10.x, 172.16-31.x, 192.168.x, 169.254.x, 0.x, localhost, IPv6 loopback/unique-local/link-local
3. `validateFetchTarget()` combines scheme + hostname check, called on every hop

13 unit tests for `isPrivateOrReservedHost` + 3 SSRF integration tests for `fetchHtml` (direct private IP, redirect to private IP, cloud metadata endpoint).

File: src/url-extractor.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    const contentLength = response.headers.get('content-length');
+    if (contentLength && parseInt(contentLength, 10) > maxLength) {
+      throw new Error(`Content too large: ${contentLength} bytes (max ${maxLength})`);
+    }
+
+    const html = await response.text();
+    if (html.length > maxLength) {
+      throw new Error(`Content too large: ${html.length} bytes (max ${maxLength})`);
+    }
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts]: Fixed in 8111c5f. Replaced `response.text()` with `readBodyWithLimit()` — a streaming reader that:

1. Reads chunks via `response.body.getReader()`
2. Accumulates `value.byteLength` per chunk
3. Throws immediately when `totalBytes > maxLength` (before buffering more)
4. Falls back to `response.text()` + `TextEncoder` byte check if `response.body` is unavailable

Content-length header check preserved as fast-path. Added test for oversized streamed body without content-length header.

File: tests/url-extractor.test.ts [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+  });
+
+  it('uses hostname as title fallback when title is missing', () => {
+    const html = `<html><head></head><body><article><h1></h1>${longParagraph}</article></body></html>`;
+    const result = extractArticle(html, 'https://myblog.example.com/post');
+
+    if (result && !result.title.includes('meaningful')) {
+      expect(result.title).toBeTruthy();
+    }
+  });
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply tests/url-extractor.test.ts:198]: Fixed in 8111c5f. All three conditional assertions converted to unconditional:

- Byline: `expect(result).not.toBeNull()` + `expect(result!.byline).toContain('John Smith')`
- Excerpt: `expect(result!.excerpt).toBeTruthy()` + length check
- Hostname fallback: `expect(result).not.toBeNull()` + `expect(result!.title).toBeTruthy()`

Also added `afterEach` cleanup for fetch mocks replacing per-test `try/finally`.

File: src/url-extractor.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  codeBlockStyle: 'fenced',
+  bulletListMarker: '-',
+});
+
+turndown.remove(['img', 'iframe', 'video', 'audio', 'picture', 'figure', 'svg']);
+
+export function isValidUrl(url: string): boolean {
+  try {
+    const parsed = new URL(url);
+    return parsed.protocol === 'http:' || parsed.protocol === 'https:';
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts:51]: Fixed in 8111c5f. Added `isPrivateOrReservedHost()` that blocks loopback, private (RFC 1918), link-local/cloud metadata (169.254.x), and IPv6 reserved ranges. `validateFetchTarget()` runs on every fetch target including redirect hops.

13 unit tests cover all blocked ranges + public hostname pass-through.

File: src/url-extractor.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    if (!contentType.includes('text/html') && !contentType.includes('text/plain') && !contentType.includes('application/xhtml')) {
+      throw new Error(`Unsupported content type: ${contentType}`);
+    }
+
+    const contentLength = response.headers.get('content-length');
+    if (contentLength && parseInt(contentLength, 10) > maxLength) {
+      throw new Error(`Content too large: ${contentLength} bytes (max ${maxLength})`);
+    }
+
+    const html = await response.text();
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts]: Fixed in 8111c5f. `readBodyWithLimit()` streams the response body via `getReader()`, counting `value.byteLength` per chunk and throwing immediately on overflow — never buffers the full body before enforcing the limit.

File: src/url-extractor.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  const maxLength = options.maxContentLength ?? DEFAULT_MAX_CONTENT_LENGTH;
+
+  const controller = new AbortController();
+  const timer = setTimeout(() => controller.abort(), timeout);
+
+  try {
+    const response = await fetch(url, {
+      signal: controller.signal,
+      headers: { 'User-Agent': USER_AGENT },
+      redirect: 'follow',
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts]: Fixed in 8111c5f. Switched to `redirect: 'manual'` with manual redirect following. Each hop is validated via `validateFetchTarget()` (scheme + hostname/IP check) before following. Max 5 redirects enforced.

File: tests/url-extractor.test.ts [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        <h1>Test</h1>
+        <span class="author" rel="author">John Smith</span>
+        ${longParagraph}
+      </article>
+    </body></html>`;
+    const result = extractArticle(html, 'https://example.com');
+
+    if (result?.byline) {
+      expect(result.byline).toContain('John Smith');
+    }
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply tests/url-extractor.test.ts]: Fixed in 8111c5f. Converted to unconditional assertion:

```ts
expect(result).not.toBeNull();
expect(result!.byline).toContain('John Smith');
```

Same treatment for excerpt and hostname fallback tests.

**Comment by mrosnerr** [comment]
## Review Feedback Addressed (8111c5f)

All actionable comments from both CodeRabbit and Cubic have been addressed. Summary:

### 🔴 P0 — SSRF Protection
- `redirect: 'manual'` with per-hop validation via `isPrivateOrReservedHost()`
- Blocks: loopback (127.x, localhost), private (10.x, 172.16-31.x, 192.168.x), link-local/cloud metadata (169.254.x), IPv6 reserved ranges
- `MAX_REDIRECTS = 5` enforced

### 🟠 P1 — Streaming Size Guard
- `readBodyWithLimit()` replaces `response.text()` — counts bytes during download, aborts immediately on overflow
- Content-length header check preserved as fast-path

### 🟠 P1 — Metadata Sanitization
- `sanitizeMetadata()` normalizes title/byline/siteName/excerpt (collapses newlines + whitespace) before markdown interpolation

### 🟡 P2 — Test Hardening
- Conditional assertions → unconditional (byline, excerpt, hostname fallback)
- `afterEach` cleanup replaces per-test `try/finally`

### ⚪ Nitpicks Addressed
- Magic number `50` → `MIN_READABLE_CHARS` named constant
- `z.string()` → `z.string().url()` for schema-layer validation
- 20 new tests added (56 total, up from 36)

### Verification
- ✅ 525 tests pass, 0 fail
- ✅ Clean `tsc` build
- ✅ 0 lint errors

**Comment by mrosnerr** [comment]
## New: Raw HTML input support (059af7e)

`knowledge-ingest` now accepts pre-fetched HTML in addition to URLs, making it composable with other MCP tools.

### What changed

**Schema** — `url` is now optional; new `html` parameter:
```typescript
{ url?: string, html?: string, model?: string }
// At least one of url or html must be provided
```

**Three usage modes:**
1. `url` only → fetches and extracts (unchanged behavior)
2. `html` only → extracts directly from provided content (no network request)
3. `url` + `html` → extracts from `html`, uses `url` for relative link resolution

**Why:** Agents with Exa, Playwright, or other web tools can now pipe pre-fetched content through `knowledge-ingest` for Readability extraction instead of re-fetching via URL.

### Also added
- `docs/tools-reference.md` — full `knowledge-ingest` documentation (was completely missing)
- `skills/open-zk-kb/SKILL.md` — agent instructions for the ingest → store workflow
- 4 new tests for html input paths

### Verification
- ✅ 529 tests pass, 0 fail
- ✅ 0 lint errors
- ✅ Clean `tsc` build

File: src/url-extractor.ts [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    if (a === 169 && b === 254) return true;            // 169.254.0.0/16 link-local / cloud metadata
+    if (a === 0) return true;                          // 0.0.0.0/8
+  }
+
+  // IPv6 bracket notation (as it appears in URL hostnames)
+  if (host.startsWith('[') && host.endsWith(']')) {
+    const ipv6 = host.slice(1, -1).toLowerCase();
+    if (ipv6 === '::1') return true;                                    // loopback
+    if (ipv6.startsWith('fc') || ipv6.startsWith('fd')) return true;    // fc00::/7 unique local
+    if (ipv6.startsWith('fe80')) return true;                           // fe80::/10 link-local
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts]: Fixed in e842b21. Replaced `startsWith('fe80')` with bitmask: `(parseInt(ipv6.slice(0,4), 16) & 0xffc0) === 0xfe80` — covers full fe80-febf range. Tests added for fe90, fea0, febf (blocked) and fec0, fe7f (allowed).

File: src/tool-handlers.ts [hunk 9]
```diff
@@ -106,6 +108,53 @@ export interface MaintainArgs {
...
+
+function sanitizeMetadata(value: string): string {
+  return value.replace(/[\r\n]+/g, ' ').replace(/\s+/g, ' ').trim();
+}
+
+export async function handleIngest(args: IngestArgs): Promise<string> {
+  let result: ExtractionResult;
+  if (args.html) {
+    const sourceUrl = args.url || 'about:blank';
+    const article = extractArticle(args.html, sourceUrl);
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/tool-handlers.ts:127]: Fixed in e842b21. Added `Buffer.byteLength(args.html, 'utf8') > MAX_HTML_BYTES` check (5MB) at the top of handleIngest before calling extractArticle. Tests added for both rejection (>5MB) and acceptance (under limit).

File: src/mcp-server.ts [hunk 10]
```diff
@@ -127,6 +127,39 @@ server.registerTool(
   },
 );
 
+// ---- knowledge-ingest ----
+
+const ingestSchema = z.object({
+  url: z.string().url().describe('URL to extract content from. Must be http:// or https://.'),
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/mcp-server.ts]: Fixed in e842b21. Added .refine() on the url field to enforce http/https at schema level. Runtime validateFetchTarget still catches it as defense-in-depth.

File: src/mcp-server.ts [hunk 11]
```diff
@@ -127,6 +127,41 @@ server.registerTool(
   },
 );
 
+// ---- knowledge-ingest ----
+
+const ingestSchema = z.object({
+  url: z.string().url().optional().describe('URL to fetch and extract content from. Provide url, html, or both (url used for link resolution when html is provided).'),
+  html: z.string().optional().describe('Raw HTML to extract content from. Use when you already fetched the page via another tool (e.g. Playwright, Exa, web_fetch).'),
+  model: z.string().optional().describe('Your model identifier (e.g. claude-opus-4, gpt-4o). Enables richer responses for capable models.'),
+});
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/mcp-server.ts]: Fixed in e842b21. Added .refine(d => d.url || d.html) on the outer schema. Runtime check in handleIngest remains as defense-in-depth.

File: src/mcp-server.ts [hunk 12]
```diff
@@ -127,6 +127,41 @@ server.registerTool(
   },
 );
 
+// ---- knowledge-ingest ----
+
+const ingestSchema = z.object({
+  url: z.string().url().optional().describe('URL to fetch and extract content from. Provide url, html, or both (url used for link resolution when html is provided).'),
+  html: z.string().optional().describe('Raw HTML to extract content from. Use when you already fetched the page via another tool (e.g. Playwright, Exa, web_fetch).'),
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/mcp-server.ts]: Fixed in e842b21 (same commit as the other size guard). Buffer.byteLength check on raw HTML input, 5MB limit matching the fetch path.

File: docs/tools-reference.md [hunk 13]
```diff
@@ -45,6 +45,59 @@ Store knowledge in the persistent knowledge base. One concept per note.
...
+| Parameter | Type | Required | Description |
+|-----------|------|----------|-------------|
+| `url` | string | No* | URL to fetch and extract content from. Must be `http://` or `https://` |
+| `html` | string | No* | Raw HTML to extract from. Use when you already fetched the page via another tool |
+
+\* At least one of `url` or `html` must be provided. When both are provided, `html` is used for extraction and `url` is used for resolving relative links.
+
+### What happens
+
+1. **If `url` only** — fetches the page with SSRF protection (private IP blocking, redirect validation, streaming size limits), then extracts
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply docs/tools-reference.md]: Fixed in e842b21. Reworded to 'literal private IP/hostname pattern blocking' and added a Security Notes section documenting what we block, the known DNS limitation, private IP link filtering, and credential stripping.

File: docs/tools-reference.md [hunk 14]
```diff
@@ -1,6 +1,6 @@
 # Tools Reference
 
-open-zk-kb exposes three MCP tools. Your AI assistant calls these automatically based on injected instructions — you rarely need to invoke them manually.
+open-zk-kb exposes four MCP tools. Your AI assistant calls these automatically based on injected instructions — you rarely need to invoke them manually.
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply docs/tools-reference.md:3]: Fixed in e842b21. README tools table updated from 'three MCP tools' to 'four MCP tools' with knowledge-ingest row.

File: src/mcp-server.ts [hunk 15]
```diff
@@ -127,6 +127,41 @@ server.registerTool(
   },
 );
 
+// ---- knowledge-ingest ----
+
+const ingestSchema = z.object({
+  url: z.string().url().optional().describe('URL to fetch and extract. Fallback only — the built-in fetcher cannot handle JavaScript-rendered pages, bot protection, or authenticated content. If you have a web tool (Playwright, Exa, web_fetch), fetch with that and pass html instead.'),
+  html: z.string().optional().describe('Preferred — raw HTML to extract content from. Pass HTML you already fetched via Playwright, Exa, web_fetch, or any browser/web tool for best results.'),
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/mcp-server.ts:137]: Fixed in e842b21. Updated url field description: 'When passing html, also pass url for relative link resolution.' Also added to docs: 'Pass url alongside html when possible — without it, relative links cannot be resolved.'

File: src/url-extractor.ts [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+/**
+ * Validates a URL for safe fetching: scheme must be http/https and
+ * hostname must not resolve to a private/reserved range.
+ */
+function validateFetchTarget(url: string): void {
+  const parsed = new URL(url);
+  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
+    throw new Error(`Blocked URL: unsupported protocol ${parsed.protocol}`);
+  }
+  if (isPrivateOrReservedHost(parsed.hostname)) {
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts:111]: Acknowledged as known limitation in e842b21. DNS resolution at the application layer isn't reliably available in Bun/Node.js without platform-specific async dns.lookup(), and fetch doesn't expose resolved IPs before connecting. Documented honestly in the new Security Notes section of tools-reference.md: 'hostnames that DNS-resolve to private IPs are not blocked — the check operates on hostname text, not resolved addresses.' Error message also updated from 'resolves to' to 'matches' for accuracy.

File: skills/open-zk-kb/SKILL.md [hunk 17]
```diff
@@ -51,6 +51,15 @@ Notes exceeding the target will trigger a soft warning — heed it and split if
 - At natural breakpoints (complex debug, architecture choice, topic change): ask *"Anything worth saving?"*
 - Before ending a session: review for uncaptured preferences, decisions, gotchas, or workflows.
 
+### Ingesting URLs
+When a useful URL comes up, use `knowledge-ingest` to extract clean article content, then store via `knowledge-store`:
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply skills/open-zk-kb/SKILL.md]: Article-like pages are the intentional scope — the tool uses Mozilla Readability which is designed for article extraction. Non-article pages (SPAs, dashboards, login forms) return null from Readability and the tool already throws 'Could not extract article content' with a clear error. The SKILL.md guidance implicitly scopes to extractable content by describing the workflow as 'extract → review → store.' A fallback to store raw URLs is already covered by knowledge-store with kind: resource.

File: skills/open-zk-kb/SKILL.md [hunk 18]
```diff
@@ -51,6 +51,18 @@ Notes exceeding the target will trigger a soft warning — heed it and split if
 - At natural breakpoints (complex debug, architecture choice, topic change): ask *"Anything worth saving?"*
 - Before ending a session: review for uncaptured preferences, decisions, gotchas, or workflows.
 
+### Ingesting URLs
+When a useful URL comes up, extract its content via `knowledge-ingest`, then store via `knowledge-store`.
+
+**If you have a web tool** (Playwright, Exa, web_fetch, dev-browser — check your available tools):
+1. Fetch the page with your web tool (handles JS rendering, bot protection, auth)
+2. `knowledge-ingest(html: "<fetched html>", url: "<source url>")` → extracts title, content, metadata
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply skills/open-zk-kb/SKILL.md:59]: Valid observation. In practice, Exa and web_fetch typically return HTML (not markdown) — the SKILL.md guidance says 'fetch with that and pass html.' Agents that use tools returning markdown would need to wrap it in basic HTML tags first, but the primary workflow (Playwright/Exa → raw HTML → knowledge-ingest) works as documented. Adding a note about this to the SKILL.md would over-complicate the instructions for an edge case.

File: src/content-splitter.ts [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    currentLines.push(line);
+  }
+
+  flush();
+  return sections;
+}
+
+const MARKDOWN_LINK_RE = /\[([^\]]*)\]\(([^)]+)\)/g;
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/content-splitter.ts:100]: Known limitation, accepted for this PR. The regex handles the vast majority of Turndown-generated markdown (which never produces parenthesized URLs or title attributes). Wikipedia-style URLs with balanced parens would need a smarter parser — tracked as future enhancement. The input is controlled (Readability → Turndown) so real-world impact is minimal.

File: src/content-splitter.ts [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+    if (/^(`{3,}|~{3,})/.test(line)) {
+      inFencedBlock = !inFencedBlock;
+      continue;
+    }
+    if (inFencedBlock) continue;
+
+    const headingMatch = line.match(/^#{2,3}\s+(.+)/);
+    if (headingMatch) {
+      currentSection = headingMatch[1].trim();
+    }
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/content-splitter.ts:121]: By design — splitSections and extractLinks both intentionally treat h4+ as body content within the parent h2/h3 section. This prevents over-fragmentation. The section label staying as the parent h2/h3 is correct behavior for link context attribution.

File: src/tool-handlers.ts [hunk 21]
```diff
@@ -106,6 +109,89 @@ export interface MaintainArgs {
...
+    if (!article) {
+      throw new Error('Could not extract article content from provided HTML. The content may not contain enough readable text.');
+    }
+    result = article;
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/tool-handlers.ts:131]: Good catch. When html is provided with an invalid url, the title fallback hits new URL(url) which could throw. The defensive try/catch on line 237 already handles this — it falls through to 'Untitled'. The error message to the agent is clear enough since extraction succeeds; the url is only used for link resolution context.

File: src/url-extractor.ts [hunk 22]
```diff
@@ -0,0 +1,255 @@
...
+      if (done) break;
+      totalBytes += value.byteLength;
+      if (totalBytes > maxLength) {
+        throw new Error(`Content too large: exceeded ${maxLength} bytes`);
+      }
+      chunks.push(value);
+    }
+  } finally {
+    reader.releaseLock();
+  }
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts:147]: Valid improvement. The reader.releaseLock() in finally lets the stream be GC'd, but explicitly canceling would be cleaner. In practice, the AbortController timeout in fetchHtml's outer scope will tear down the connection. Accepted as minor tech debt — will address in a follow-up.

File: src/content-splitter.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/content-splitter.ts:100]: Same as CodeRabbit's comment on this line — the regex is intentionally simple because the input is controlled (Readability → Turndown output). Turndown never produces image markdown (we call turndown.remove(['img', ...])). Parenthesized URLs are a known limitation accepted for this PR — the vast majority of extracted links work correctly. Tracked as future enhancement.

**Comment by mrosnerr** [comment]
## All review feedback addressed (e842b21)

This commit addresses all outstanding review comments from both CodeRabbit and Cubic:

### 🔴 Security fixes
- **fe80::/10 IPv6 range** — bitmask check now covers full fe80-febf range (6 new tests)
- **HTML size guard** — 5MB `Buffer.byteLength` check on raw HTML input path (2 new tests)
- **Redirect URL provenance** — `fetchHtml` returns `finalUrl` after redirect chain; `extractArticle` uses the final URL for correct link resolution and dedupe (2 new tests)
- **Extracted link filtering** — private IP links stripped from output via `canonicalizeUrl` (3 new tests)
- **URL credential stripping** — `url.username`/`url.password` cleared before output (1 new test)

### 🟡 Schema & validation
- **Zod schema** — `.refine()` enforces http/https on url + requires at least one of url/html
- **Error message** — "resolves to" → "matches" (no DNS resolution claim)
- **Docstring** — updated to match

### 📝 Docs
- **tools-reference.md** — new Security Notes section with honest DNS limitation caveat
- **README** — knowledge-ingest added to tools table
- **Schema description** — mentions passing url alongside html for link resolution

### 🧹 Code quality
- `countWords` deduplicated (single export from content-splitter)
- `about:blank` title fallback → `'Untitled'` instead of empty string
- Word count from `article.textContent` (not markdown syntax)
- `extractLinks` prefix check: `http://`/`https://` instead of `http`

**663 tests pass, 0 failures, 0 lint errors.** 26 new tests added for all changed behavior.

@coderabbitai @cubic-dev-ai please re-review.

File: src/url-extractor.ts [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+    return false;
+  }
+}
+
+/**
+ * Returns true if the hostname matches a private, loopback, link-local,
+ * or otherwise reserved IP pattern. Does not perform DNS resolution.
+ */
+export function isPrivateOrReservedHost(hostname: string): boolean {
+  const host = hostname.toLowerCase();
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts]: Fixed in 36ac490. Added `.replace(/\\.\$/, '')` to strip trailing dots before hostname matching. Tests added for `localhost.`, `app.localhost.` (blocked) and `example.com.` (allowed).

File: src/url-extractor.ts [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+
+  if (!result) {
+    throw new Error(`Could not extract article content from ${url}. The page may not contain enough readable content.`);
+  }
+
+  logToFile('DEBUG', 'URL extraction complete', {
+    url: finalUrl,
+    title: result.title,
+    wordCount: result.wordCount,
+  });
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/url-extractor.ts]: Fixed in 36ac490. Credentials are now stripped from finalUrl before passing to extractArticle and logToFile — new URL(finalUrl) then clear username/password. Test added verifying admin:secret credentials don't appear in ExtractionResult.url.

File: src/mcp-server.ts [hunk 25]
```diff
@@ -127,6 +127,44 @@ server.registerTool(
...
+}).refine(d => d.url || d.html, { message: 'At least one of url or html must be provided' });
+
+server.registerTool(
+  'knowledge-ingest',
+  {
+    description: 'Extract article content as clean markdown. Returns title, content, word count, and metadata. PREFER passing html from your own web tools (Playwright, Exa, web_fetch) — the built-in url fetcher is a basic fallback that cannot render JavaScript or bypass bot protection. Use the extracted content to create notes via knowledge-store.',
+    inputSchema: ingestSchema as unknown as AnySchema,
+  },
+  async (args: z.infer<typeof ingestSchema>) => {
+    try {
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/mcp-server.ts:148]: Fixed in 36ac490. Replaced try/catch pattern with `await expect(fetchHtml(...)).rejects.toThrow(/matches a private\/reserved range/)` — test now fails properly if fetchHtml stops throwing.

File: src/storage/NoteRepository.ts [hunk 26]
```diff
@@ -839,6 +839,19 @@ export class NoteRepository {
...
   }
 
+  findByUrl(url: string): Array<{ id: string; title: string }> {
+    try {
+      const parsed = new URL(url);
+      const hostPath = parsed.hostname + parsed.pathname.replace(/\/$/, '');
+      const stmt = this.db.prepare(
+        "SELECT id, title FROM notes WHERE content LIKE ? AND status != 'archived' LIMIT 5"
+      );
+      return stmt.all(`%${hostPath}%`) as Array<{ id: string; title: string }>;
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/storage/NoteRepository.ts:850]: Fixed in 36ac490. hostPath now escapes \\, %, and _ before LIKE interpolation, and the query uses ESCAPE '\\' clause. Percent-encoded URL paths can no longer widen the match.

File: src/storage/NoteRepository.ts [hunk 27]
```diff
@@ -839,6 +839,19 @@ export class NoteRepository {
...
+    } catch {
+      return [];
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply src/storage/NoteRepository.ts:854]: Fixed in 36ac490. Same as Cubic's comment — LIKE metacharacters escaped with ESCAPE clause.

File: tests/url-extractor.test.ts [hunk 28]
```diff
@@ -0,0 +1,973 @@
...
+describe('SSRF error message clarity', () => {
+  it('says matches not resolves', async () => {
+    try {
+      await fetchHtml('http://127.0.0.1/secret');
+    } catch (e: unknown) {
+      expect((e as Error).message).toContain('matches a private/reserved range');
+      expect((e as Error).message).not.toContain('resolves to');
+    }
+  });
+});
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply tests/url-extractor.test.ts:990]: Fixed in 36ac490. Same as Cubic's comment — replaced try/catch with rejects.toThrow.
